### ISSUE TITLE
Support: Language Injection with EmmyLua `@language LANG`

### ISSUE THREAD

**Issue by captainko** [MunifTanjim/tree-sitter-lua#4]
I want to implement [EmmyLua Language Injection](https://emmylua.github.io/annotations/language.html), but I don't know how to achieve this with the syntax tree.
Please let me know if you know how to do it.

```lua
---@language JSON
local jsonText = [[{
    "name":"Emmy"
}]]
```

**Comment by MunifTanjim** [comment]
This is as close as you can get:

```scheme
(
  (comment) @language
  .
  (variable_declaration
    (assignment_statement
      (expression_list
        (string
          content: _ @content))))
)
```

The value of `@language` is
```
---@language JSON
```
and the value of `@content` is
```
{
    "name":"Emmy"
}
```

So you'll need a custom [predicate](https://tree-sitter.github.io/tree-sitter/using-parsers#predicates) to detect that it's JSON.

**Comment by MunifTanjim** [comment]
In case you wanna hardcode it for `JSON`, then you can do:

```lua
(
  (comment) @_annotation
  .
  (variable_declaration
    (assignment_statement
      (expression_list
        (string
          content: _ @json))))
  (#eq? @_annotation "---@language JSON")
)
```

**Comment by captainko** [comment]
Thank you for such detailed instruction. But could we achieve this in a more robust way like [tree-sitter language injection](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection)?

**Comment by MunifTanjim** [comment]
> But could we achieve this in a more robust way like tree-sitter language injection

Not sure what you meant by "robust". As I understand, that link just provides a guideline. Different editors can implement it in different ways. So it depends on which editor you're targeting and how that editor implements it.

**Comment by captainko** [comment]
Sorry for making my point clear enough. I meant how we can inject languages dynamically without hardcoding the pattern for each language.
I take the example from tree-sitter document.

Consider this ruby code:
```ruby
system <<-BASH.strip!
  abc --def | ghi > jkl
BASH
```
With this syntax tree:
```
(program
  (method_call
    method: (identifier)
    arguments: (argument_list
      (call
        receiver: (heredoc_beginning)
        method: (identifier))))
  (heredoc_body
    (heredoc_end)))
```
The following query would specify that the contents of the heredoc should be parsed using a language named “BASH” (because that is the text of the heredoc_end node):

```
(heredoc_body
  (heredoc_end) @injection.language) @injection.content
```
As you can see in the example (heredoc_end) is actually `BASH` and the `@inject.content` is marked as `bash`. 

If we can do this we can inject any language as we like.

**Comment by MunifTanjim** [comment]
Got it. It's as I said in the first comment: https://github.com/MunifTanjim/tree-sitter-lua/issues/4#issuecomment-1024166872

The whole `---@language JSON` is a single node `comment`. So in order to extract the language name from there you'll need a custom [predicate](https://tree-sitter.github.io/tree-sitter/using-parsers#predicates). For example, something like this:

```
(
  (comment) @injection.language
  .
  (variable_declaration
    (assignment_statement
      (expression_list
        (string
          content: _ @injection.content))))
  (#custom-predicate-to-remove-first-n-characters! @injection.language 13)
)
```

So, the implementation of the `#custom-predicate-to-remove-first-n-characters!` predicate needs to remove the first 13 characters (i.e. `---@language `) from the `@injection.language` capture. Or something similar to this.

In short, unless the environment where tree-sitter is running (e.g. [nvim-treesitter/nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter)) has support for a custom predicate that can manipulate the captured text, you can't do it dynamically (for this parser).
